File: tests/test_argon2.cpp
#include <gtest/gtest.h>
#include "Argon2Hasher.hpp"
#include <vector>
#include <string>
#include <stdexcept>
#include <chrono>
#include <iomanip>
#include <functional>
#include <cstring>

using namespace mylib::crypto;

class Argon2HasherTest : public ::testing::Test {
protected:
    

    std::string makeStoredHash(const std::string& password) {
        return Argon2Hasher::hashPasswordWithSalt(password);
    }
};

TEST_F(Argon2HasherTest, HashPasswordWithSalt_ReturnsValidFormat) {
    auto result = Argon2Hasher::hashPasswordWithSalt("testPassword");
    EXPECT_NE(result.find('$'), std::string::npos);
    
    auto [salt, hash] = Argon2Hasher::splitSaltAndHash(result);
    EXPECT_FALSE(salt.empty());
    EXPECT_FALSE(hash.empty());
}

TEST_F(Argon2HasherTest, HashPasswordWithSalt_GeneratesDifferentSalts) {
    auto result1 = Argon2Hasher::hashPasswordWithSalt("testPassword");
    auto result2 = Argon2Hasher::hashPasswordWithSalt("testPassword");
    EXPECT_NE(result1, result2);
}





class Argon2PerformanceTest : public ::testing::Test {
protected:
    struct BenchmarkResult {
        double avgTimeMs;
        double opsPerSec;
        std::string config;
    };
    
    std::vector<BenchmarkResult> results;
    
    void runBenchmark(const std::string& name, 
                     std::function<void()> func,
                     int runs = 5) {
        // Прогрев гоев
        for (int i = 0; i < 2; ++i) func();
        
        auto start = std::chrono::high_resolution_clock::now();
        for (int i = 0; i < runs; ++i) func();
        auto end = std::chrono::high_resolution_clock::now();
        
        auto totalMs = std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count();
        double avgMs = static_cast<double>(totalMs) / runs;
        double opsPerSec = 1000.0 / avgMs * runs;
        
        results.push_back({avgMs, opsPerSec, name});
        
        std::cout << std::fixed << std::setprecision(2);
        std::cout << "[PERF] " << std::setw(25) << std::left << name 
                  << " " << std::setw(8) << avgMs << " ms/op"
                  << " " << std::setw(10) << opsPerSec << " ops/sec\n";
    }
    
    void TearDown() override {
        if (!results.empty()) {
            std::cout << "\nPerformance Summary:\n";
            std::cout << "-----------------------------------------\n";
            for (const auto& r : results) {
                std::cout << std::setw(25) << std::left << r.config 
                          << " " << std::setw(8) << r.avgTimeMs
                          << " " << std::setw(10) << r.opsPerSec << "\n";
            }
            std::cout << "-----------------------------------------\n";
        }
    }
};

TEST_F(Argon2PerformanceTest, HashPerformance) {
    runBenchmark("Default params", []() {
        char pwd[] = "performanceTest";
        auto hash = Argon2Hasher::hashPasswordWithSalt(pwd);
        std::memset(pwd, 0, sizeof("performanceTest"));
    });
    
    runBenchmark("High t_cost", []() {
        char pwd[] = "performanceTest";
        auto hash = Argon2Hasher::hashPasswordWithSalt(pwd, 5);
        std::memset(pwd, 0, sizeof("performanceTest"));
    });
}

TEST_F(Argon2PerformanceTest, VerifyPerformance) {
    const auto stored = Argon2Hasher::hashPasswordWithSalt("testPassword");
    
    runBenchmark("Verify correct", [&stored]() {
        char pwd[] = "testPassword";
        bool result = Argon2Hasher::verifyPassword(pwd, stored);
        std::memset(pwd, 0, sizeof("testPassword"));
    });
    
    runBenchmark("Verify wrong", [&stored]() {
        char pwd[] = "wrongPassword";
        bool result = Argon2Hasher::verifyPassword(pwd, stored);
        std::memset(pwd, 0, sizeof("wrongPassword"));
    });
}

using namespace std;

TEST_F(Argon2PerformanceTest, PerformanceDifferentParameters) {
    const int iterations = 10; 
    const vector<tuple<string, int, size_t>> test_params = {
        {"Fast (t=1, m=16MB)", 1, 1<<16},      
        {"Balanced (t=3, m=64MB)", 3, 1<<18},  
        {"Secure (t=5, m=256MB)", 5, 1<<20}   
    };

    cout << "\nArgon2 Performance results (avg per operation):\n";
    cout << "----------------------------------------------------------------------------\n";
    cout << "| Config                 | Hash (ms) | Verify (ms) | Memory (MB) | Threads |\n";
    cout << "----------------------------------------------------------------------------\n";

    const string test_password = "testPassword123";

    for (const auto& [name, t_cost, m_cost] : test_params) {

        auto start_hash = chrono::high_resolution_clock::now();
        string stored_hash;
        for (int i = 0; i < iterations; ++i) {
            stored_hash = Argon2Hasher::hashPasswordWithSalt(test_password, t_cost, m_cost);
        }
        auto end_hash = chrono::high_resolution_clock::now();
        double hash_time = chrono::duration_cast<chrono::milliseconds>(end_hash - start_hash).count() / iterations;


        auto start_verify = chrono::high_resolution_clock::now();
        for (int i = 0; i < iterations; ++i) {
            bool result = Argon2Hasher::verifyPassword(test_password, stored_hash);
        }
        auto end_verify = chrono::high_resolution_clock::now();
        double verify_time = chrono::duration_cast<chrono::milliseconds>(end_verify - start_verify).count() / iterations;

        cout << "| " << setw(22) << left << name << " | "
             << setw(9) << fixed << setprecision(2) << hash_time << " | "
             << setw(11) << verify_time << " | "
             << setw(11) << (m_cost/(1024*1024)) << " | "
             << setw(7) << "1" << " |\n";
    }
    cout << "----------------------------------------------------------------------------\n";


    cout << "\nArgon2 Performance with different password lengths (t=3, m=64MB, p=1):\n";
    cout << "--------------------------------------------------------\n";
    cout << "| Password Length | Hash (ms) | Verify (ms) |\n";
    cout << "--------------------------------------------------------\n";

    const vector<pair<string, size_t>> password_lengths = {
        {"8 chars", 8},
        {"32 chars", 32},
        {"1KB", 1024},
        {"64KB", 65536}
    };

    for (const auto& [desc, length] : password_lengths) {
        string password(length, 'a');
        
        auto start_hash = chrono::high_resolution_clock::now();
        string stored_hash;
        for (int i = 0; i < iterations; ++i) {
            stored_hash = Argon2Hasher::hashPasswordWithSalt(password, 3, 1<<18);
        }
        auto end_hash = chrono::high_resolution_clock::now();
        double hash_time = chrono::duration_cast<chrono::milliseconds>(end_hash - start_hash).count() / iterations;

        auto start_verify = chrono::high_resolution_clock::now();
        for (int i = 0; i < iterations; ++i) {
            bool result = Argon2Hasher::verifyPassword(password, stored_hash);
        }
        auto end_verify = chrono::high_resolution_clock::now();
        double verify_time = chrono::duration_cast<chrono::milliseconds>(end_verify - start_verify).count() / iterations;

        cout << "| " << setw(16) << desc << " | "
             << setw(9) << hash_time << " | "
             << setw(11) << verify_time << " |\n";
    }
    cout << "--------------------------------------------------------\n";
}

File: testsPython/testArgon2.py
import pytest
from ShSlibPy import Argon2Hasher
import time
from typing import Tuple
import re

# Unit тесты для Argon2
class TestArgon2Basic:
    def test_hash_password_with_salt_format(self):
        """Проверка формата возвращаемого хеша"""
        result = Argon2Hasher.hashPasswordWithSalt("testPassword")
        assert '$' in result
        
        parts = result.split('$')
        assert len(parts) == 2
        assert len(parts[0]) > 0  # salt
        assert len(parts[1]) > 0  # hash

    def test_hash_password_with_salt_unique_salts(self):
        """Проверка генерации разных солей для одного пароля"""
        result1 = Argon2Hasher.hashPasswordWithSalt("testPassword")
        result2 = Argon2Hasher.hashPasswordWithSalt("testPassword")
        assert result1 != result2

    def test_verify_password_correct(self):
        """Проверка верификации правильного пароля"""
        password = "correctPassword"
        stored_hash = Argon2Hasher.hashPasswordWithSalt(password)
        assert Argon2Hasher.verifyPassword(password, stored_hash)

    def test_verify_password_incorrect(self):
        """Проверка верификации неправильного пароля"""
        password = "correctPassword"
        wrong_password = "wrongPassword"
        stored_hash = Argon2Hasher.hashPasswordWithSalt(password)
        assert not Argon2Hasher.verifyPassword(wrong_password, stored_hash)

    def test_verify_password_tampered_hash(self):
        """Проверка верификации с измененным хешем"""
        password = "testPassword"
        stored_hash = Argon2Hasher.hashPasswordWithSalt(password)
        
        # Изменяем хеш
        parts = stored_hash.split('$')
        tampered_hash = f"{parts[0]}${parts[1][:-1]}x"  # меняем последний символ
        
        assert not Argon2Hasher.verifyPassword(password, tampered_hash)

    def test_custom_parameters(self):
        """Проверка работы с кастомными параметрами"""
        password = "testPassword"
        stored_hash = Argon2Hasher.hashPasswordWithSalt(
            password, 
            t_cost=5, 
            m_cost=65536,  # 64MB
            salt_len=32,
            out_len=64
        )
        
        # Проверяем формат
        parts = stored_hash.split('$')
        assert len(parts[0]) == 32 * 2  # salt в hex
        assert len(parts[1]) == 64 * 2   # hash в hex
        
        # Проверяем верификацию
        assert Argon2Hasher.verifyPassword(
            password, 
            stored_hash,
            t_cost=5,
            m_cost=65536,
            out_len=64
        )

# Тесты производительности Argon2
class TestArgon2Performance:
    @pytest.fixture(autouse=True)
    def setup(self):
        self.test_password = "testPassword123"
        self.iterations = 5

    def measure_time(self, func, iterations=1):
        """Измеряет среднее время выполнения функции"""
        # Прогрев
        for _ in range(2):
            func()
        
        start = time.perf_counter()
        for _ in range(iterations):
            func()
        end = time.perf_counter()
        
        return (end - start) / iterations

    def test_hash_performance_default_params(self):
        """Тест производительности хеширования с параметрами по умолчанию"""
        def hash_func():
            return Argon2Hasher.hashPasswordWithSalt(self.test_password)
        
        avg_time = self.measure_time(hash_func, self.iterations)
        print(f"\n[PERF] Hash (default): {avg_time:.4f} sec")
        
        # Проверяем, что время в разумных пределах (менее 1 сек)
        assert avg_time < 1.0

    def test_verify_performance(self):
        """Тест производительности верификации"""
        stored_hash = Argon2Hasher.hashPasswordWithSalt(self.test_password)
        
        def verify_correct():
            return Argon2Hasher.verifyPassword(self.test_password, stored_hash)
        
        def verify_incorrect():
            return Argon2Hasher.verifyPassword("wrongPassword", stored_hash)
        
        correct_time = self.measure_time(verify_correct, self.iterations)
        incorrect_time = self.measure_time(verify_incorrect, self.iterations)
        
        print(f"[PERF] Verify (correct): {correct_time:.4f} sec")
        print(f"[PERF] Verify (incorrect): {incorrect_time:.4f} sec")
        
        assert correct_time < 1.0
        assert incorrect_time < 1.0

    def test_performance_with_different_parameters(self):
        """Тест производительности с разными параметрами"""
        test_params = [
            ("Fast (t=1, m=16MB)", 1, 1<<16),      # 16MB
            ("Balanced (t=3, m=64MB)", 3, 1<<18),   # 64MB
            ("Secure (t=5, m=256MB)", 5, 1<<20)     # 256MB
        ]
        
        print("\nArgon2 Performance with different parameters:")
        print("| Config                 | Hash (sec) | Verify (sec) |")
        print("|------------------------|------------|--------------|")
        
        for name, t_cost, m_cost in test_params:
            # Измеряем время хеширования
            def hash_func():
                return Argon2Hasher.hashPasswordWithSalt(
                    self.test_password, 
                    t_cost=t_cost, 
                    m_cost=m_cost
                )
            
            hash_time = self.measure_time(hash_func, self.iterations)
            
            # Измеряем время верификации
            stored_hash = hash_func()
            
            def verify_func():
                return Argon2Hasher.verifyPassword(
                    self.test_password, 
                    stored_hash,
                    t_cost=t_cost,
                    m_cost=m_cost
                )
            
            verify_time = self.measure_time(verify_func, self.iterations)
            
            print(f"| {name:<22} | {hash_time:>10.4f} | {verify_time:>12.4f} |")
        
        print("|------------------------|------------|--------------|")

    def test_performance_with_different_password_lengths(self):
        """Тест производительности с паролями разной длины"""
        password_lengths = [
            ("8 chars", "a" * 8),
            ("32 chars", "a" * 32),
            ("1KB", "a" * 1024),
            ("64KB", "a" * 65536)
        ]
        
        print("\nArgon2 Performance with different password lengths (t=3, m=64MB):")
        print("| Password Length | Hash (sec) | Verify (sec) |")
        print("|-----------------|------------|--------------|")
        
        for name, password in password_lengths:
            # Измеряем время хеширования
            def hash_func():
                return Argon2Hasher.hashPasswordWithSalt(
                    password, 
                    t_cost=3, 
                    m_cost=1<<18  # 64MB
                )
            
            hash_time = self.measure_time(hash_func, self.iterations)
            
            # Измеряем время верификации
            stored_hash = hash_func()
            
            def verify_func():
                return Argon2Hasher.verifyPassword(
                    password, 
                    stored_hash,
                    t_cost=3,
                    m_cost=1<<18
                )
            
            verify_time = self.measure_time(verify_func, self.iterations)
            
            print(f"| {name:<15} | {hash_time:>10.4f} | {verify_time:>12.4f} |")
        
        print("|-----------------|------------|--------------|")

    def test_thread_safety(self):
        """Проверка потокобезопасности (если применимо)"""
        from concurrent.futures import ThreadPoolExecutor
        
        stored_hash = Argon2Hasher.hashPasswordWithSalt(self.test_password)
        
        def verify_task(_):
            return Argon2Hasher.verifyPassword(self.test_password, stored_hash)
        
        with ThreadPoolExecutor(max_workers=4) as executor:
            results = list(executor.map(verify_task, range(10)))
        
        assert all(results)
